### PR TITLE
Strikeout deprecated CompletionItems

### LINKED ISSUE #50972: Strikeout CompletionItems (and Symbols?) that are deprecated
<kbd>Ctrl</kbd>+<kbd>Space</kbd> is nice and all but when you are new to a library you just see a list of functions and you won't realize that some of them are deprecated. Would it be possible to add a strikeout line via the `line-through` `text-decoration` or some kind of icon overlay to indicate that something is deprecated?

image

### PR DESCRIPTION
**Pull request by kamranayub**
Closes #50972 

## Changes

- Add `CompletionItem.deprecated`
- Add `deprecated` flag to `SuggestDataDto`
- Add inline styling to suggest widget

### CODE REVIEW

**Comment by kamranayub** [comment]
@jrieken I wasn't sure if the way I was applying styles was recommended so definitely looking for feedback there (and whether or not to externalize the CSS class name)

File: src/vs/editor/contrib/suggest/suggestWidget.ts [hunk 1]
```diff
@@ -9,6 +9,7 @@ import { createMatches } from 'vs/base/common/filters';
 import * as strings from 'vs/base/common/strings';
 import { Event, Emitter } from 'vs/base/common/event';
 import { onUnexpectedError } from 'vs/base/common/errors';
+import { ClassName } from 'vs/editor/common/model/intervalTree';
```
Comment (jrieken) [inline src/vs/editor/contrib/suggest/suggestWidget.ts]: This dependency shouldn't exist. Define a suggest-widget-private style and simply refer to its classname further down.

**Review by jrieken — CHANGES_REQUESTED** [review]

File: src/vs/editor/contrib/suggest/suggestWidget.ts (hunk 1 above)
Comment (jrieken) [reply src/vs/editor/contrib/suggest/suggestWidget.ts]: :top: I guess that answers https://github.com/microsoft/vscode/pull/78092#issuecomment-515993758. No need to externalise/re-use

File: src/vs/editor/contrib/suggest/suggestWidget.ts [hunk 2]
```diff
@@ -62,6 +62,10 @@ export const editorSuggestWidgetForeground = registerColor('editorSuggestWidget.
 export const editorSuggestWidgetSelectedBackground = registerColor('editorSuggestWidget.selectedBackground', { dark: listFocusBackground, light: listFocusBackground, hc: listFocusBackground }, nls.localize('editorSuggestWidgetSelectedBackground', 'Background color of the selected entry in the suggest widget.'));
 export const editorSuggestWidgetHighlightForeground = registerColor('editorSuggestWidget.highlightForeground', { dark: listHighlightForeground, light: listHighlightForeground, hc: listHighlightForeground }, nls.localize('editorSuggestWidgetHighlightForeground', 'Color of the match highlights in the suggest widget.'));
 
+/**
+ * Suggest widget styles
+ */
+const editorSuggestWidgetDeprecatedClassName = "suggest-widget-deprecated";
```
Comment (kamranayub) [inline src/vs/editor/contrib/suggest/suggestWidget.ts]: I am getting a `Unexternalized string found: "suggest-widget-deprecated"` on this line from hygiene. Advice?

Comment (jrieken) [reply src/vs/editor/contrib/suggest/suggestWidget.ts]: We use double quotes for nls string (natural language strings, needs externalisation for translation). Use single quotes for 'other' string.

**Comment by KamasamaK** [comment]
Note that adding a dedicated `deprecated` property would conflict with @jrieken's proposal at 
https://github.com/microsoft/vscode/issues/23927#issuecomment-387807469

**Comment by kamranayub** [comment]
@KamasamaK Confused though, because I thought it was already defined in the LSP:

https://microsoft.github.io/language-server-protocol/specification#textDocument_completion

Won't this change match that interface defined there? I assumed the interface is a 1:1 mapping.

**Comment by jrieken** [comment]
> Won't this change match that interface defined there? I assumed the interface is a 1:1 mapping.

There is no 1:1 mapping, tho often a conceptual mapping between the extension API and the LSP. @KamasamaK raises a valid point here, not sure how to proceed. Adding a single boolean is obviously the easiest change but we don't wanna do that all the time, e.g. for private, static, readonly, etc, etc. So maybe some upfront design work is needed. Also checking what @dbaeumer's plans for LSP are?

**Comment by kamranayub** [comment]
Without a lot of background/context for this, my uneducated view is that `private`, `static`, `readonly` would be [access modifiers](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/access-modifiers) while `deprecated` seems like a separate concern, like an annotation/tag/status.

**Comment by jrieken** [comment]
> , my uneducated view is that private, static, readonly would be access modifiers while deprecated seems like a separate concern

Well, we don't wanna be to academic here. Sure 'public', 'private', 'protected', 'friend' etc are access modifiers but there is more, like 'static' or 'final' or Java's multithread modifiers like 'volatile' and 'transient'. I think there is no absolute truth and the API never tried to correctly represent actual language constructs but to be a common denominator of a UI model the reflects those things. So, one can think of a bit mask of such flags that add some extra rendering to a symbol, like a *static method*, or a *private, deprecated constructor*

**Comment by dbaeumer** [comment]
I would support adding a modifier instead of a separate `deprecated` field. There is no rule for a 1:1 mapping. And there are requests in LSP to support more than `deprecated` which would lead to a modifier as well.

**Comment by kamranayub** [comment]
Once I have a hot second I can explore switching to a modifier. 👍

**Comment by jrieken** [comment]
I will go ahead here and merge things and do more the remaining changes on master. Thanks @kamranayub

**Comment by kamranayub** [comment]
Awesome, thanks! 🙏

**Comment by jrieken** [comment]
No - thank you!

**Pull request merged**